### DISCUSSION TITLE
A new Enviroment works fine with AutoComplete but not in MainMenu.

### DISCUSSION THREAD

**Discussion by rafa4mat** [WinEdt-Team/WinEdt-Support#87] [1.1 General Help and Discussion]
I´m adding wrapfigure enviroment

I created

    "c:\Users\...\AppData\Roaming\WinEdt Team\WinEdt 11\Macros\Recorder\TeXWrapfigure.edt" 

modifying a copy of TeXFigure.edt

    // Added
    LetReg(-9,'%?');
    LetTempReg;
    
    LetReg(-10, '\begin{%-9}{}%\');
    LetReg(-10,'%-0%!?%-1%% You can add width {75mm} AND even insert [n lines] between \begin{wrapfigure} and  {...} %-2%\');
    LetReg(-10, '%-0%!?%-1\centering   %% You can add \unitlegth=1mm %\');
    LetReg(-10, '%-0%!?%-1%\');
    //LetReg(-10, '%-0%!?%-1\includegraphics[width=]{}   %% You can use tikzpicture or \picture instead \includegraphics%\');
    LetReg(-10, '%-0%!?%-1\caption{}\label{}%\');
    LetReg(-10, '%-0%!?\end{%-9}');
    
    // This will be inserted by the AutoPopup Exit Event Handler:
    LetTempReg("%-0");
    
    End;

It works fine adding in AutoComplete:

    OBJECT="wrapfigure"
          MACRO="LetReg(-1,!'  ');"+           // Indentation
                "LetReg(-2,!'');"+             // Floating placement eg. [htbp]
                "Exe('%!h\WinEdt Team\WinEdt 11\Macros\Recorder\TeXWrapfigure.edt');"

I added previously  in Enviroment.ini (using Options Interface)

    ENVIRONMENT="Wraptable" // RSF Added 2024-12-22
      MODE_FILTER="TeX"
      ALLOW_SOFT_WRAP=1
      SCOPE=1
      PRIORITY=5
      STRICT_PRIORITY=0
      CASE_SENSITIVE=1
      START="\begin{wraptable}"
      STOP="\end{wraptable}"
      FORMAT=-1
      LINE_WRAP=-1
      AUTO_INDENT=0
      FREE_CARET=0
      INPUT_TABS=0

All works fine BUT...

...BUT no works (no error but no code is writted) in Main Menu (and later I´ would try in Tools.ini)

I added in MainMenu.ini (usgin OptionsInterface):

        ITEM="Figure"
          CAPTION="&Figure"
          IMAGE="Figure"
          MACRO="Exe('%b\Menus\Insert\Image.edt');"
          REQ_DOCUMENT=1
        ITEM="Wrapfigure"
          CAPTION="&Wrapfigure"
          IMAGE="Figure"
          MACRO="Exe('%!h\WinEdt Team\WinEdt 11\Macros\Recorder\TeXWrapfigure.edt');"
          REQ_DOCUMENT=1  

Any help?.

**Comment by Alex-Simonic** [comment]
Your macro TeXFigure.edt is written to be executed inside the Auto-Completion event handler. It assigns a string to the temporary register %? but by itself, this does not insert anything in your document. The code:

```
// This will be inserted by the AutoPopup Exit Event Handler:
LetTempReg("%-0");
```

should be replaced with something like:

```
Mark(0);
Ins("%-0");
Goto(0);
NextBullet;
```
Furthermore, the initial value of temporary registers %? and %!? will be undefined when executing such macro from the menu (autocomplete event handler initializes these values before the macro is called). You will have to take care of that (assign the proper values) in a macro executed from the menu.

In other words, the macro works fine from the menu: it assigns the registers but that is all that this code does by itself. Now you know...

**Comment by rafa4mat** [comment]
Thank you very much, Alex. Furthermore, reading your comment I realized that

one  .edt is at ...Menus/insert/... 
and 
the other (another different file)  is at ... /Macros/Auto Complete/...

diferent folders and different code, as you explained.


I keep trying to add Menu items...

Thank you, Alex.
